### PR TITLE
IGNITE-22581 Add planner test to verify numeric type coercion of source for INSERT, UPDATE and MERGE operators

### PR DESCRIPTION
**Pull request by ygerzhedovich**
https://issues.apache.org/jira/browse/IGNITE-22581

### CODE REVIEW

File: modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java [hunk 1]
```diff
@@ -351,4 +355,50 @@ public static RexNode generateLiteralOrValueExpr(ColumnType type, Object value)
...
+                return SqlTypeName.DATE;
+            case TIME:
+                return SqlTypeName.TIME;
+            case DATETIME:
+                return SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE;
+            case TIMESTAMP:
+                return SqlTypeName.TIMESTAMP;
+            case UUID:
+            case BITMASK:
+                return SqlTypeName.OTHER;
```
Comment (lowka) [inline modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java:391]: There is no type that uses SqlTypeName.OTHER. UUID uses ANY, and there is no SqlType for BITMASK.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningPredicateSelfTest.java [hunk 2]
```diff
@@ -118,12 +120,20 @@ public void testLiteralValue(ColumnType columnType) {
         expectPartitionsPruned(table, columns, new Object[0], group, val);
     }
 
+    private static NativeType getNativeType(ColumnType columnType) {
+        SqlTypeName sqlTypeName = SqlTestUtils.columnType2SqlTypeName(columnType);
+        int precision = IgniteTypeSystem.INSTANCE.getMaxPrecision(sqlTypeName);
+        int scale = IgniteTypeSystem.INSTANCE.getMaxScale(sqlTypeName);
+
+        return TypeUtils.columnType2NativeType(columnType, precision, scale, 4);
```
Comment (lowka) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningPredicateSelfTest.java:128]: The length parameter should have the same value as the precision (because they are the same thing).

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java [hunk 3]
```diff
@@ -499,8 +499,6 @@ protected <T extends RelNode> void assertPlan(
         String planString = RelOptUtil.dumpPlan("", plan, SqlExplainFormat.TEXT, DEFAULT_EXPLAIN_LEVEL);
         log.info("statement: {}\n{}", sql, planString);
 
-        checkSplitAndSerialization(plan, schemas);
```
Comment (lowka) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:502]: I would rather introduce an option to disable. this check with a default value that tells to always check.

Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:502]: planner tests shouldn't check serialization

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningPredicateSelfTest.java (hunk 2 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningPredicateSelfTest.java:128]: agree. Fixed

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java (hunk 3 above)
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:502]: How are going to ensure that a plan can be sent to another node?

File: modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java (hunk 1 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java:391]: fixed for UUID. For BITMASK it doesn't matter, we don't support the type and it should be get rid under IGNITE-22283

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+        assertPlan("INSERT INTO T VALUES(" + val + "," + val + ")", schema, keyValOperandMatcher(operandMatcher)::matches, List.of());
+    }
+
+    @ParameterizedTest
+    @MethodSource("argsDyn")
+    public void insertDynamicParameters(
+            TypePair pair,
+            Matcher<RexNode> operandMatcher
+    ) throws Exception {
+        // ToDo: remove during implement IGNITE-22283
```
Comment (lowka) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java:71]: Do we really what to introduce another style of writing `TODO` ?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java (hunk 3 above)
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:502]: KeyValue plans are not serialisable, but it does not mean that all serialisation should be disabled.

Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:502]: This check greatly simplifies detection bugs like this one https://issues.apache.org/jira/browse/IGNITE-19858

Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:502]: partially covered by our integration tests. If we need wider cover it should be separated tests, not as part of planner tests

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java (hunk 4 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java]: Always use the same approach. But anyway - amended to ``TODO``

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java (hunk 3 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java]: reworked a bit

**Review by lowka — APPROVED** [review]

File: modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java [hunk 5]
```diff
@@ -351,4 +355,50 @@ public static RexNode generateLiteralOrValueExpr(ColumnType type, Object value)
...
+            case NUMBER:
+                return SqlTypeName.DECIMAL;
+            case DATE:
+                return SqlTypeName.DATE;
+            case TIME:
+                return SqlTypeName.TIME;
+            case DATETIME:
+                return SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE;
+            case TIMESTAMP:
+                return SqlTypeName.TIMESTAMP;
```
Comment (korlov42) [inline modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java:388]: vice versa

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/BaseTypeCoercionTest.java [hunk 6]
```diff
@@ -51,6 +53,14 @@ static Stream<Arguments> allNumericPairs() {
         return Arrays.stream(NumericPair.values()).map(Arguments::of);
     }
 
+    static void checkIncludesAllTypePairs(Stream<Arguments> args) {
```
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/BaseTypeCoercionTest.java:56]: it's better to revert this change. BaseTypeCoercionTest is base class for all type pairs, not only numeric ones.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java [hunk 7]
```diff
@@ -675,7 +676,16 @@ protected void checkSplitAndSerialization(IgniteRel rel, IgniteSchema publicSche
         checkSplitAndSerialization(rel, Collections.singleton(publicSchema));
     }
 
+    // Set of Relational operators that do not support serialization and shouldn't be sent between cluster nodes.
+    private static final Set<Class> unsupportSerializationOperators = Set.of(
+            IgniteKeyValueModify.class
```
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/AbstractPlannerTest.java:681]: IgniteKeyValueGet as well

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * A set of test to verify behavior of type coercion for INSERT operations, when values belongs to the NUMERIC type family.
+ *
+ * <p>This tests aim to help to understand in which cases implicit cast will be added to which values.
+ */
+public class InsertSourcesCoercionTest extends BaseTypeCoercionTest {
```
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java:49]: let's add `Numeric-` prefix to conform to other test classes inside the package. The same for other classes

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/UpdateSourcesCoercionTest.java [hunk 9]
```diff
@@ -0,0 +1,687 @@
...
+/**
+ * A set of test to verify behavior of type coercion for UPDATE operations, when values belongs to the NUMERIC type family.
+ *
+ * <p>This tests aim to help to understand in which cases implicit cast will be added to which values.
+ */
+public class UpdateSourcesCoercionTest extends BaseTypeCoercionTest {
+
+    @ParameterizedTest
+    @MethodSource("args1")
+    public void update1(
```
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/UpdateSourcesCoercionTest.java:54]: let's give it meaningful name. The same for `update2` method

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/UpdateSourcesCoercionTest.java [hunk 10]
```diff
@@ -0,0 +1,687 @@
...
+
+            @Override
+            public void describeTo(Description description) {
+
+            }
+        };
+    }
+
+
+    private static Stream<Arguments> args1() {
```
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/UpdateSourcesCoercionTest.java:134]: the same, let's give it meaningful name

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/BaseTypeCoercionTest.java (hunk 6 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/BaseTypeCoercionTest.java:56]: I will just rename the method

File: modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java (hunk 5 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java]: fixed

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java (hunk 8 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java]: doen

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/UpdateSourcesCoercionTest.java (hunk 9 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/UpdateSourcesCoercionTest.java]: done

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java (hunk 8 above)
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/InsertSourcesCoercionTest.java]: you forgot to update `MergeSourcesCoercionTest`

**Review by korlov42 — APPROVED** [review]
